src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,221 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size = 0;
+    private Node<T> first;
+    private Node<T> last;
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            first = new Node<>(null, value, null);
+            last = first;
+        } else {
+            last.next = new Node<>(last, value, null);
+            last = last.next;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexValidityToAdd(index);
+
+        if (index == 0 && size > 0) {
+            addToTheBeginning(value);
+        }
+        if (index == size) {
+            add(value);
+        } else if (index > 0) {
+            addToTheMiddle(value, index);
+        }
+    }
+
+    private void checkIndexValidityToAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    private void addToTheMiddle(T value, int index) {
+        if (index < size / 2) {
+            int i = 0;
+            for (MyLinkedList.Node<T> x = first; x != null; x = x.next) {
+                if (i == index) {
+                    createNodeAndItsLinks(value, x);
+                    break;
+                }
+                i++;
+            }
+        } else {
+            int i = size - 1;
+            for (MyLinkedList.Node<T> x = last; x != null; x = x.prev) {
+                if (i == index) {
+                    createNodeAndItsLinks(value, x);
+                    break;
+                }
+                i--;
+            }
+        }
+        size++;
+    }
+
+    private void createNodeAndItsLinks(T value, Node<T> x) {
+        Node<T> newNode = new Node<>(x.prev, value, x);
+        x.prev.next = newNode;
+        x.prev = newNode;
+    }
```

Comment:
Be careful with linking Nodes. You are creating a new Node and linking it with the previous Node, but you are not linking it with the next Node. Make sure that you add mutual links from both sides when you insert a new Node.